src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Unix/PkcsPal.Unix.cs

```diff
@@ -3,17 +3,209 @@
 // See the LICENSE file in the project root for more information.
 
 using System;
+using System.Buffers;
 using System.Diagnostics;
+using System.Security.Cryptography;
+using System.Security.Cryptography.Asn1;
+using System.Security.Cryptography.Pkcs;
+using System.Security.Cryptography.X509Certificates;
 
 namespace Internal.Cryptography
 {
     internal abstract partial class PkcsPal
     {
-        private static PkcsPal s_instance;
+        private static PkcsPal s_instance = new ManagedPkcsPal();
 
-        static PkcsPal()
+        private class ManagedPkcsPal : PkcsPal
         {
-            throw new PlatformNotSupportedException();
+            public override byte[] Encrypt(
+                CmsRecipientCollection recipients,
+                ContentInfo contentInfo,
+                AlgorithmIdentifier contentEncryptionAlgorithm,
+                X509Certificate2Collection originatorCerts,
+                CryptographicAttributeObjectCollection unprotectedAttributes)
+            {
+                throw new PlatformNotSupportedException(SR.PlatformNotSupported_CryptographyPkcs);
+            }
+
+            public override DecryptorPal Decode(
+                byte[] encodedMessage,
+                out int version,
+                out ContentInfo contentInfo,
+                out AlgorithmIdentifier contentEncryptionAlgorithm,
+                out X509Certificate2Collection originatorCerts,
+                out CryptographicAttributeObjectCollection unprotectedAttributes)
+            {
+                throw new PlatformNotSupportedException(SR.PlatformNotSupported_CryptographyPkcs);
+            }
+
+            public override byte[] EncodeOctetString(byte[] octets)
+            {
+                AsnWriter writer = new AsnWriter(AsnEncodingRules.DER);
+                writer.WriteOctetString(octets);
+                return writer.Encode();
+            }
+
+            public override byte[] DecodeOctetString(byte[] encodedOctets)
+            {
+                AsnReader reader = new AsnReader(encodedOctets, AsnEncodingRules.BER);
```

Comment:
Could you explain why the decoder uses BER but the encoder (line 44) uses DER?

Reply:
The spec says that the data is BER encoded, so we use a BER reader.  All DER is BER, so writing DER is acceptable.  Since some readers may assume DER (or some of DER's rules) because it's easier than BER writing DER just helps everyone.

But I'll add some comments, too :smile: